Timer provides #getMinTime() method

lembergerth opened the pull request:
To get the minimum measured time across all intervals.

It would be interesting for me to also get the minimum time that some repeated computation took,
but was this left out on purpose?

lembergerth committed:
Timer provides #getMinTime() method

To get the minimum measured time across all intervals.

PhilippWendler commented:
The minimum was left out because it was not yet interesting so far, and keeping track of it costs something. What is your use case?

Does it make sense to count the current running interval for minimum time? E.g., if one starts the timer and immediately asks for minimum, it will return a low number that might not be near the actual minimum.

PhilippWendler commented on src/org/sosy_lab/common/time/Timer.java:
```diff
@@ -183,10 +189,22 @@ public TimeSpan getMaxTime() {
   }
 
+  long minTime() {
+    return Math.min(minTime, currentInterval());
```
This is broken. It always returns 0 except if the timer is running.

lembergerth commented:
> The minimum was left out because it was not yet interesting so far, and keeping track of it costs something. What is your use case?

I want to get a feeling for how fast two (mostly) independent computations can be.
While the maximum value is interesting, I'd also like to know whether it takes "at least 5 seconds",
or "at least 0.5 seconds".

> Does it make sense to count the current running interval for minimum time? E.g., if one starts the timer and immediately asks for minimum, it will return a low number that might not be near the actual minimum.

I don't have that use case, but just wanted to be consistent with max value.
If you don't think this is useful, I'll simply remove the corresponding methods.

PhilippWendler commented:
We can add it, but what should it return while the timer is running?

PhilippWendler commented on src/org/sosy_lab/common/time/Timer.java:
```diff
@@ -85,6 +85,11 @@
+  /**
+   * The minimal time of all intervals.
+   */
+  private long minTime = 0;
```
This is also wrong: minTime will always stay zero.

PhilippWendler commented:
Oh, and what should the result be before the first start of the timer?

lembergerth committed:
Minimum timer value is properly computed.

Minimum timer value is not always 0, but is initially Long.MAX_VALUE
so that it is computed correctly. On access,
minimum value 0 is returned if the timer still has the initial value.

Running intervals are ignored, because we do not know
yet whether they will be smaller or larger than the minimum.

lembergerth commented:
In my opinion, a running timer should not be considered because it is not sure yet whether the value will be the minimum or not.
Can you think of a case where the user may want the timer to include a still running computation as the minimum? I imagine that that would be confusing and lead to wrong conclusions, but maybe I'm missing something...

Before the first start of the timer, I would return 0, because no time was consumed.
This would be similar to the maximum value of 0.

Both changes are incorporated into the last commit.
One could also keep `minTime = 0` and check in `stop()` whether the current intervall that is stopped
is the first interval, but I thought that the initial value of `MAX_VALUE` would be less error-prone.

PhilippWendler commented:
Ok.

Could you please run `ant format-source`?

lembergerth committed:
Format Timer.java
